serving/docker/partition/trt_llm_partition.py

```diff
@@ -1,39 +1,190 @@
+            )
+        if args.cluster_key is not None:
+            cluster_config = dict(cluster_key=args.cluster_key)
+        else:
+            cluster_config = infer_cluster_config()
+
+        # This should only be used for debugging.
+        # The env var BUILDER_FORCE_NUM_PROFILES should override the number of
+        # optimization profiles during TRT build.
+        # BUILDER_FORCE_NUM_PROFILES must be less than or equal to the number of
+        # optimization profiles set by model's prepare_inputs().
+        force_num_profiles_from_env = os.environ.get(
+            "BUILDER_FORCE_NUM_PROFILES", None)
+        if force_num_profiles_from_env is not None:
+            logger.warning(
+                f"Overriding # of builder profiles <= {force_num_profiles_from_env}."
+            )
 
+        build_config = BuildConfig.from_dict(
+            {
+                'max_input_len': args.max_input_len,
+                'max_seq_len': args.max_seq_len,
+                'max_batch_size': args.max_batch_size,
+                'max_beam_width': args.max_beam_width,
+                'max_num_tokens': args.max_num_tokens,
+                'opt_num_tokens': args.opt_num_tokens,
+                'max_prompt_embedding_table_size':
+                args.max_prompt_embedding_table_size,
+                'gather_context_logits': args.gather_context_logits,
+                'gather_generation_logits': args.gather_generation_logits,
+                'strongly_typed': True,
+                'force_num_profiles': force_num_profiles_from_env,
+                'weight_sparsity': args.weight_sparsity,
+                'profiling_verbosity': args.profiling_verbosity,
+                'enable_debug_output': args.enable_debug_output,
+                'max_draft_len': args.max_draft_len,
+                'speculative_decoding_mode': speculative_decoding_mode,
+                'input_timing_cache': args.input_timing_cache,
+                'output_timing_cache': args.output_timing_cache,
+                'auto_parallel_config': {
+                    'world_size':
+                    args.auto_parallel,
+                    'gpus_per_node':
+                    args.gpus_per_node,
+                    'sharded_io_allowlist': [
+                        'past_key_value_\\d+',
+                        'present_key_value_\\d*',
+                    ],
+                    'same_buffer_io': {
+                        'past_key_value_(\\d+)': 'present_key_value_\\1',
+                    },
+                    **cluster_config,
+                },
+                'dry_run': args.dry_run,
+                'visualize_network': args.visualize_network,
+                'max_encoder_input_len': args.max_encoder_input_len,
+                'weight_streaming': args.weight_streaming,
+                'monitor_memory': args.monitor_memory,
+            },
+            plugin_config=plugin_config)
 
-def main():
```

Comment:
Better to keep `main()`, variables defined in `if __name__ == "__main__":` considered as global variables, they might conflict with parameters passed into `build_engine()`, PyCharm will give warnings.

Reply:
updated